### PR TITLE
[clang-format] Change LLVM style to BreakAfterAttributes == ABS_Leave

### PR DESCRIPTION
**Pull request by Endilll**
This patch addresses some examples of bad formatting in Clang. The following commit contains only changes suggested by clang-format: https://github.com/llvm/llvm-project/pull/70349/commits/21689b56d1fc1db0b2263e8049ff656d3757ad36. I believe it's a net negative on readability, with a couple of particularly bad cases. Highlights:
```diff
-    [[clang::preferred_type(bool)]]
-    mutable unsigned CachedLocalOrUnnamed : 1;
+    [[clang::preferred_type(bool)]] mutable unsigned CachedLocalOrUnnamed : 1;
```
```diff
-    [[clang::preferred_type(TypeDependence)]]
-    unsigned Dependence : llvm::BitWidth<TypeDependence>;
+    [[clang::preferred_type(TypeDependence)]] unsigned Dependence
+        : llvm::BitWidth<TypeDependence>;
```
```diff
-    [[clang::preferred_type(ExceptionSpecificationType)]]
-    unsigned ExceptionSpecType : 4;
+    [[clang::preferred_type(
+        ExceptionSpecificationType)]] unsigned ExceptionSpecType : 4;
```

### CODE REVIEW

**Comment by AaronBallman** [comment]
I'm in favor of these changes. The current formatting makes some pretty bad decisions in places, but I think it's a hard problem to solve -- some attributes are longer and may want to be on their own line, other attributes are shorter and maybe are reasonable to inline, but there will always be the ones in the middle where it could go either way and will be wrong periodically. Picking "leave them where the developer put them" seems like a pretty reasonable compromise to me.

Uncertain how others feel about it though.

**Comment by owenca** [comment]
This would change the default for every other predefined style, which might be ok if there's no mention of breaking after C++11 attributes in those styles.

**Comment by Endilll** [comment]
[Mozilla](https://firefox-source-docs.mozilla.org/code-quality/coding-style/coding_style_cpp.html), [WebKit](https://webkit.org/code-style-guidelines/), [GCC](https://gcc.gnu.org/codingconventions.html#Cxx_Conventions), [Microsoft](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions), and [Chromium](https://chromium.googlesource.com/chromium/src/+/main/styleguide/c++/c++.md) doesn't appear to be opinionated on attributes at all.

[Google](https://google.github.io/styleguide/cppguide.html) is the only one mentioning attributes at all:
> Attributes, and macros that expand to attributes, appear at the very beginning of the function declaration or definition, before the return type:
>```cpp
>  ABSL_ATTRIBUTE_NOINLINE void ExpensiveFunction();
>  [[nodiscard]] bool IsOk();
>```

I'm not sure which `BreakAfterAttributes` option this boils down to, but I don't mind setting it back to `Never` for Google style if reviewers would like me to do so.

**Comment by owenca** [comment]
> [Google](https://google.github.io/styleguide/cppguide.html) is the only one mentioning attributes:
> 
> > Attributes, and macros that expand to attributes, appear at the very beginning of the function declaration or definition, before the return type:
> > ```c++
> >  ABSL_ATTRIBUTE_NOINLINE void ExpensiveFunction();
> >  [[nodiscard]] bool IsOk();
> > ```
> 
> I'm not sure which `BreakAfterAttributes` option this boils down to, but I don't mind setting it back to `Never` for Google style if reviewers would like me to do so.

I think `Leave` would work.

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -794,6 +794,7 @@ clang-format
 ------------
 - Add ``AllowBreakBeforeNoexceptSpecifier`` option.
 - Add ``AllowShortCompoundRequirementOnASingleLine`` option.
+- Change ``BreakAfterAttributes`` from ``Never`` to ``Leave`` in LLVM style
```
Comment (owenca) [inline clang/docs/ReleaseNotes.rst:797]: Full stop at the end of the sentence.

File: clang/unittests/Format/FormatTest.cpp [hunk 2]
```diff
@@ -26179,7 +26179,7 @@ TEST_F(FormatTest, RemoveSemicolon) {
 
 TEST_F(FormatTest, BreakAfterAttributes) {
   FormatStyle Style = getLLVMStyle();
-  EXPECT_EQ(Style.BreakAfterAttributes, FormatStyle::ABS_Never);
+  EXPECT_EQ(Style.BreakAfterAttributes, FormatStyle::ABS_Leave);
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:26182]: Can you move it down to just above line 26196, with an empty line separating it from `Code` above?
```
                           "  i = 0;\n"
                           "  return 1;\n"
                           "}");

  EXPECT_EQ(Style.BreakAfterAttributes, FormatStyle::ABS_Leave);
  verifyNoChange(Code, Style);

  Style.BreakAfterAttributes = FormatStyle::ABS_Never;
  verifyFormat("[[nodiscard]] inline int f(int &i);\n"
               "[[foo([[]])]] [[nodiscard]] int g(int &i);\n"
               "[[nodiscard]] inline int f(int &i) {\n"
```

**Review by owenca — APPROVED** [review]

File: clang/unittests/Format/FormatTest.cpp (hunk 2 above)
Comment (Endilll) [reply clang/unittests/Format/FormatTest.cpp]: Done!

File: clang/docs/ReleaseNotes.rst (hunk 1 above)
Comment (Endilll) [reply clang/docs/ReleaseNotes.rst]: Fixed.

**Review by owenca — APPROVED** [review]

**Review by mydeveloperday — APPROVED** [review]

**Pull request merged**